### PR TITLE
noto-fonts: 2020-01-23 -> unstable-2022-04-25 and other changes

### LINKED ISSUE #166953: Noto-fonts: switch to variable fonts
### Describe the bug

This is a follow up to https://github.com/NixOS/nixpkgs/pull/166600 and the comment in https://github.com/NixOS/nixpkgs/blob/a0a77f1c738384eb82560b9d359e27afb35c4dfc/pkgs/data/fonts/noto-fonts/default.nix#L34

AFAICT, just like the variable fonts are best for CJK, they are the best version to use for all the other Noto fonts too. This is what [Fedora thinks](https://fedoraproject.org/wiki/Changes/VariableNotoFonts#Benefit_to_Fedora). The variable fonts save a lot of disk space (10.94MiB vs 1.02MiB for Sinhala) and provide all variations, as well as enabling custom font variations in software that supports them.

I was wondering about hinting but it seems that all that's [happening](https://github.com/googlefonts/noto-source/blob/main/scripts/notobuilder.py#L46) is [ttfautohint](https://github.com/googlefonts/gftools/blob/0ede0eb58feeb0b44ad9ef9fb0ea7987a9b2b3ef/Lib/gftools/builder/autohint.py#L97). ttfautohint is built into FreeType so I think it gets run on Linux regardless for TTF fonts.

Now, the Adobe CFF renderer for OTF is better than the TTF renderer, and ideally we would get OTF variable fonts like with CJK. But those aren't available at the moment.

unhinted/slim-variable-ttf is a variable font with only 400 (Regular) through 700 (Bold) on the weight axis, prepared for Android system UI use ([script](https://github.com/googlefonts/noto-source/blob/main/misc_scripts/slimmer-vfs.sh#L20), [doc note](https://github.com/notofonts/noto-docs/issues/3)), it's smaller at 353 KB vs 1.03MB. Since we ship Light (300) in the main noto-fonts package I think we want the full range. Although Fedora [chose](https://src.fedoraproject.org/rpms/google-noto-fonts/blob/rawhide/f/google-noto-fonts.spec#_940) the slim, the savings from removing weights seems small IMO.

I'm opening this now but I think we should wait a week or two before switching over the fonts, in case there is more fallout from the variable CJK switch.

### Notify maintainers
cc @midchildan since you've been writing all the patches recently

### PR DESCRIPTION
**Pull request by Mathnerd314**
###### Description of changes

* noto-tools: adjust dependencies to allow running notodiff
* noto-fonts: 
  * 2020-01-23 -> unstable-2022-04-25; various fonts have been added and updated
  * switch to variable fonts (closes #166953)
  * split out Chrome OS core fonts (Arimo, Tinos, Cousine) to new output croscore-fonts
  * add Emoji to test:
image

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] Tested NixOS module test `noto-fonts.tests.noto-fonts`
- [x] Tested basic functionality of notodiff
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking - the worst case for the croscore change is falling back to an ugly font so I don't think it's necessary
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/data/fonts/noto-fonts/default.nix [hunk 1]
```diff
@@ -103,15 +125,11 @@ let
...
-    weights = "{Regular,Bold,Light,Italic,BoldItalic,LightItalic}";
-  };
-
-  noto-fonts-extra = mkNoto {
-    pname = "noto-fonts-extra";
-    weights = "{Black,Condensed,Extra,Medium,Semi,Thin}*";
-  };
+  noto-fonts = noto-pkg.out;
+  noto-fonts-extra = noto-pkg.extra;
+  croscore-fonts = noto-pkg.croscore;
```
[the remark this answers is not in the document]
Comment (sternenseemann) [reply pkgs/data/fonts/noto-fonts/default.nix:130]: Alternatively, set `meta.outputsToInstall`.

[the remark this answers is not in the document]
Comment (Mathnerd314) [reply pkgs/data/fonts/noto-fonts/default.nix:130]: I see that nix-env is buggy, but who will actually use nix-env? Installing fonts with nix-env doesn't add them to the system font list, you would have to use a hacky workaround like https://discourse.nixos.org/t/how-to-install-fonts-on-a-non-nixos-system/14758.

[the remark this answers is not in the document]
Comment (Mathnerd314) [reply pkgs/data/fonts/noto-fonts/default.nix:130]: And outputsToInstall is `[ "out" ]` so only the base noto fonts will be installed, the hacks will only be needed for croscore and extra.

[the remark this answers is not in the document]
Comment (midchildan) [reply pkgs/data/fonts/noto-fonts/default.nix:130]: I think fonts installed by nix-env would be picked up by fontconfig on NixOS systems. The default configuration in the fontconfig package includes the Nix user profile and default profile in its search path.

https://github.com/NixOS/nixpkgs/blob/b4cc9cd38f05f1a764e21bfb1b14e89be76068b0/pkgs/development/libraries/fontconfig/make-fonts-conf.xsl#L38-L48

[the remark this answers is not in the document]
Comment (midchildan) [reply pkgs/data/fonts/noto-fonts/default.nix:130]: Also for non-NixOS systems, all solutions involve adding `~/.nix-profile/share/fonts` to the fontconfig search path in one way or another so avoiding nix-env wouldn't help in this particular case.

[the remark this answers is not in the document]
Comment (sternenseemann) [reply pkgs/data/fonts/noto-fonts/default.nix:130]: > the hacks will only be needed for croscore and extra.

I believe `nix-env` doesn't install any additional outputs if an output is selected explicitly which would be the case here.

File: pkgs/data/fonts/noto-fonts/default.nix [hunk 2]
```diff
@@ -11,32 +11,44 @@
...
-        sha256 = "0zc1r7zph62qmvzxqfflsprazjf6x1qnwc2ma27kyzh6v36gaykw";
+        rev = "a19de47f845dbd4c61b884c7ff90ce993555d05d";
+        sparseCheckout = ''
+          unhinted/otf
+          unhinted/variable-ttf
+        '';
+        hash = "sha256-7Jzo7402CiOcn2q++zbAzn2tP/y2x7QJuwjcP1NoyLQ=";
       };
 
+      analyzeScan = ./analyze_scan.py;
```
Comment (sternenseemann) [inline pkgs/data/fonts/noto-fonts/default.nix:35]: Can you add a comment what this does roughly?

**Review by sternenseemann — COMMENTED** [review]
I'm not sure if there's a simpler solution than `analyzeScan`, but I think it's fine to merge. Whatever works for the maintainer is great :)

File: pkgs/data/fonts/noto-fonts/default.nix (hunk 2 above)
Comment (Mathnerd314) [reply pkgs/data/fonts/noto-fonts/default.nix:35]: It reads all the font information from fc-scan and figures out which fonts are included in the variable fonts and which aren't and need the OTF version.

Comment (Mathnerd314) [reply pkgs/data/fonts/noto-fonts/default.nix:35]: There probably is a way to do it with shell scripting but my head hurt thinking about it.

**Comment by Mathnerd314** [comment]
I was wondering if these fonts would have an issue similar to #171976, but I tested it and it seems these variable fonts work fine in Discord.

File: pkgs/data/fonts/noto-fonts/default.nix [hunk 3]
```diff
@@ -11,32 +11,48 @@
...
-        #
-        # TODO: install OpenType, variable versions?
-        local out_ttf=$out/share/fonts/truetype/noto
-        install -m444 -Dt $out_ttf phaseIII_only/unhinted/ttf/*/*-${weights}.ttf
-        install -m444 -Dt $out_ttf phaseIII_only/hinted/ttf/*/*-${weights}.ttf
-        install -m444 -Dt $out_ttf unhinted/*/*-${weights}.ttf
-        install -m444 -Dt $out_ttf hinted/*/*-${weights}.ttf
+        fc-scan ./unhinted -f '"%{fullname[0]}","%{file}"\n' > scan.csv
+        python $analyzeScan
+        local out_ttf=$out/share/fonts/google-noto
```
Comment (sternenseemann) [inline pkgs/data/fonts/noto-fonts/default.nix:44]: Fonts are no longer installed to the `truetype` subdirectory. This seems like it would break at least `koreader`.

Comment (Mathnerd314) [reply pkgs/data/fonts/noto-fonts/default.nix:44]: Freetype scans all subdirectories regardless of name so the directory names are arbitrary. koreader [uses Freetype](https://github.com/koreader/crengine/blob/master/crengine/src/lvfntman.cpp#L5011) so should work fine. Certainly Fedora doesn't bother with the truetype directory anymore, see e.g. [Droid sans](https://koji.fedoraproject.org/koji/rpminfo?rpmID=30381402). Also it would be misleading in this case to call them "truetype" since they are a mixture of opentype (otf) and variable TTFs.

The more questionable aspect is whether koreader supports variable fonts. It seems to have the individual TTF file names names [hard-coded](https://github.com/koreader/koreader/blob/master/frontend/ui/font.lua#L39).  When I run koreader it crashes with the message ` frontend/ui/font.lua:353: attempt to index local 'face' (a nil value)`. This could be because it doesn't like the variable fonts I have installed.

Comment (sternenseemann) [reply pkgs/data/fonts/noto-fonts/default.nix:44]: Sorry, my comment was more (or less?) concrete than that: The `koreader` derivation assumes this at the moment which should be easy to fix:

https://github.com/NixOS/nixpkgs/blob/fbdc30b994138a9e82fdb08b21781ab45fd81ef3/pkgs/applications/misc/koreader/default.nix#L46-L48

Comment (Mathnerd314) [reply pkgs/data/fonts/noto-fonts/default.nix:44]: Ah, I see. koreader should probably fetch the noto fonts itself in a private derivation, like [OnlyOffice](https://github.com/NixOS/nixpkgs/blob/634141959076a8ab69ca2cca0f266852256d79ee/pkgs/applications/office/onlyoffice-bin/default.nix#L57). Particularly it needs only a few languages and font variants, as compared to this package. cc @contrun and @neonfuz who can write that patch. (koreader crashes for me so I don't think I can do it)

Comment (sternenseemann) [reply pkgs/data/fonts/noto-fonts/default.nix:44]: @a-m-joseph Any chance you could look into this, seeing as you recently touched `koreader`?

**Comment by a-m-joseph** [comment]
Small request: could you please put the version bump in a commit by itself that does nothing else?

Ideally in the first or last commit of the series, but if that isn't possible then in the middle is fine.

This will make things simpler for anybody with local customizations that need to be adapted as a result of this PR merging.  They will have an easier time bisecting/cherrypicking.

File: pkgs/data/fonts/noto-fonts/default.nix (hunk 3 above)
Comment (a-m-joseph) [reply pkgs/data/fonts/noto-fonts/default.nix:44]: > The more questionable aspect is whether koreader supports variable fonts. It seems to have the individual TTF file names names [hard-coded](https://github.com/koreader/koreader/blob/master/frontend/ui/font.lua#L39). When I run koreader it crashes with the message ` frontend/ui/font.lua:353: attempt to index local 'face' (a nil value)`. This could be because it doesn't like the variable fonts I have installed.

No, that means it couldn't find the fonts.

I get the same error above with this PR.

If I revert this PR I don't get the error.

> Ah, I see. koreader should probably fetch the noto fonts itself in a private derivation

Yeah, I'm not excited about this but I don't see any other way considering the hardcoded font names in koreader.  That project already has an astonishing level of vendoring going on.  I will write the PR to remove the `koreader`->`noto-fonts` dependency and have `koreader` do its own separate `fetchurl` to get the fonts.  I'm leaving town tomorrow morning and if I try to get it done before then I will end up doing a sloppy/rushed job.  This will need to wait until mid/late next week.  Feel free to ping me or open a bug and assign it to me as a reminder.

**Comment by Mathnerd314** [comment]
> Small request: could you please put the version bump in a commit by itself that does nothing else?

They changed the directory layout between the revisions (the directory `phaseIII_only` is gone), so I don't see a way to make a version update work without including the new font selection script. And that's pretty much f4cad5a54855ad872a5e7ca38df0b5d1de8c8f53, with a few testing changes thrown in.

> I'm leaving town tomorrow morning and if I try to get it done before then I will end up doing a sloppy/rushed job. This will need to wait until mid/late next week. 

I ended up working most of this out, https://github.com/Mathnerd314/nixpkgs/commit/1ca433ac84cf5dec3a06f19ad2f61fdb56ba5175. The resulting package even seems to work on my system (opening the default help / tutorial thing), yay! It still says:
```
06/02/22-12:04:57 ERROR failed to register crengine font: cannot register font <./fonts/noto/NotoSerif-BoldItalic.ttf>
06/02/22-12:04:57 ERROR failed to register crengine font: cannot register font <./fonts/noto/NotoSerif-Regular.ttf>
```
though.

But there's no rush, this PR is pretty much stalled while the various variable font issues work themselves out.

File: pkgs/data/fonts/noto-fonts/default.nix (hunk 3 above)
Comment (a-m-joseph) [reply pkgs/data/fonts/noto-fonts/default.nix:44]: > Unfortunately just convincing their CMake machinery to use pre-downloaded tarballs (instead of insisting on downloading them itself) has been so difficult

I've finally made some real progress on this.  Hope to have a draft PR shortly.

Comment (a-m-joseph) [reply pkgs/data/fonts/noto-fonts/default.nix:44]: > This seems like it would break at least `koreader`.

TL;DR: this problem disappears when we [we build koreader from source](https://github.com/NixOS/nixpkgs/pull/178320) instead of trying to rebundle the Debian binaries.

It turns out that nixpkgs' `koreader/default.nix` having a dependency upon nixpkgs' `noto-fonts` is 100% an artifact of the fact that `koreader/default.nix` is not much more than `dpkg -x`.  The problem disappears as soon as we build koreader from source.

Koreader has a [git submodule](https://github.com/koreader/koreader/tree/master/resources) containing the exact font files they expect (since they hardwire so many things, like the fonts' filenames).

These fonts are, for some reason, *not* included in the `.deb` archives that our current `koreader/default.nix` is using.  I suppose this could be in anticipation of Debian policy, although koreader has never been carried by Debian and would need a lot of other more drastic changes in order for them to accept it.  So I really don't know why the fonts get dropped when the `deb` is built.

In any event: my built-from-source `koreader` solves the problem here.

**Comment by Luflosi** [comment]
What's missing to get this merged?

**Comment by sternenseemann** [comment]
Figuring out what to do about koreader: Is it already broken? Will it be broken by this change? Can we fix it?

**Comment by amjoseph-nixpkgs** [comment]
> Figuring out what to do about koreader: Is it already broken? Will it be broken by this change? Can we fix it?

The situation with koreader is largely a result of the fact that by using binary packages we are downstream of Debian and inherit their packaging choices.

Upstream pins and vendors the exact fonts they need, as a git submodule, and when you build from source those wind up in the output.

Debian undoes this as part of their packaging.

Then we re-do it as part of our repackaging, but we don't re-do it exactly as upstream intended: instead of putting in the pinned+vendored fonts that upstream intends, we swap in our own `noto-fonts` package.  So when our `noto-fonts` is bumped we're out of sync with what upstream intends.

**Comment by Luflosi** [comment]
To me it sounds line the `koreader` package needs to be fixed to pin the correct version of `noto-fonts`.
Otherwise we will never be able to update `noto-fonts`.

**Comment by sternenseemann** [comment]
We can also opt to mark it as broken and take care of it later.
